t.rast.line: new addon to draw trendlines based on raster layers from a strds

ecodiv opened the pull request:
New addon to draw trendlines based on the raster layers from a spatio-temporal raster database, showing for each time step the average values of the raster layer. If a zonal map is provided, trend lines are drawn showing for each time step the average value of each of the categories of the zonal raster layer.

ecodiv committed:
t.rast.line: new addon, trendlines per raster category

Draws a trendline based on the raster layers in a strds. Optionally, separate trendlines are drawn per category of a input cover layer.

ecodiv committed:
work in progress

ecodiv committed:
version 0.9

ecodiv committed:
option to set line width enabled

Se the line with of the grid lines to half that of the line with of the trend lines and alpha 0.5. For now hard coded. Possibly add an option later to change this.

ecodiv committed:
Merge branch 'OSGeo:grass8' into t_rast_line

ecodiv committed:
t.rast.line addon

New addon to draw trendlines based on the raster layers from a spatio-temporal raster database, showing for each time step the average values of the raster layer. If a zonal map is provided, trend lines are drawn showing for each time step the average value of each of the categories of the zonal raster layer.

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+<h2>DESCRIPTION</h2>
+
+<em>t.rast.line</em> draws trend lines of the average values of the 
+input raster layers in a space-time raster database (strd). The trend 
```
```suggestion
input raster layers in a space-time raster dataset (strds). The trend 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<em>t.rast.line</em> draws trend lines of the average values of the 
+input raster layers in a space-time raster database (strd). The trend 
+line represents the average values of the whole computational region. 
```
```suggestion
line represents the average values of the current computational region. 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+<em>t.rast.line</em> draws trend lines of the average values of the 
+input raster layers in a space-time raster database (strd). The trend 
+line represents the average values of the whole computational region. 
+The user can optionally opt to show an error bar for each trend line 
```
```suggestion
The user can optionally show an error bar for each trend line 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+can select a subset of the raster layers using the WHERE conditions.
+
+<p>
+By default, the resulting plot is displayed on screen. However, the 
+user can also save the plot to file using the <b>output</b> option. The 
```
```suggestion
By default, the resulting plot is displayed on a new screen. However, the 
user can also save the plot to a file using the <b>output</b> option. The 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+user can also save the plot to file using the <b>output</b> option. The 
+format is determined by the extension given by the user. So, if output 
+= outputfile.png, the plot will be saved as a *png* file. The user can 
+set the output size (in inches) and resolution (dpi) of the output 
+image.
```
```suggestion
set the output size (in inches) and resolution (dpi).
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<h2>NOTE</h2>
+
+The user can specify the number of threads to be used with the nprocs 
```
```suggestion
The user can specify the number of threads to be used with the <b>nprocs</b> 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<p>
+First step is to create temporal datasets <i>tempmean</i> and 
+<i>precip_sum</i> for respectively the rainfall and temperature time 
+series, as described in <a 
```
```suggestion
<i>precip_sum</i> for the rainfall and temperature time 
series respectively, as described in <a 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<h2>EXAMPLE</h2>
+
+The next two examples use the North Carolina sample data (NC) and North 
```
```suggestion
The next two examples use the North Carolina full (NC) and North 
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<p>
+<div class="code"><pre>
+g.region raster=2000_01_tempmean@climate_2000_2012
```
```suggestion
g.region raster=2000_01_tempmean
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+<p>
+<div class="code"><pre>
+g.region raster=2000_01_tempmean@climate_2000_2012
+t.rast.line input=tempmean@climate_2000_2012 nprocs=10
```
```suggestion
t.rast.line input=tempmean nprocs=10
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<p>
+<div class="code"><pre>
+t.rast.line input=precip_sum@climate_2000_2012 \
+    error=sd line_color=0:128:0:255 nprocs=10
```
```suggestion
t.rast.line input=precip_sum error=sd line_color=0:128:0:255 nprocs=10
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+19 = 2 Brunswick
+* = NULL" > recl.txt
+
+r.reclass input=boundary_county_500m@PERMANENT \
+    output=comparison_counties rules=recl.txt --overwrite
+
+t.rast.line -l input=tempmean@climate_2000_2012 \
+    zones=comparison_counties error=sd nprocs=10
```
```suggestion
r.reclass input=boundary_county_500m output=comparison_counties rules=recl.txt

t.rast.line -l input=tempmean zones=comparison_counties error=sd nprocs=10
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+# Create the plot
+
+t.rast.line -l -g input=tempmean@climate_2000_2012 \
+    zones=comparison_counties error=sd nprocs=10
```
```suggestion
t.rast.line -l -g input=tempmean zones=comparison_counties error=sd nprocs=10
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<h3>Example 5</h3>
+
+You can zoom in on a specific period using the <b>where</b> option. For example, to plot the trend line for the time period 01-01-2004 to 01-01-2010, you can use the following:
```
```suggestion
You can zoom in on a specific period using the <b>where</b> option. 
For example, to plot the trend line for the time period 01-01-2004 to 
01-01-2010, you can use the following:
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<p>
+<div class="code"><pre>
+t.rast.line -l -g input=tempmean@climate_2000_2012 \
+    zones=comparison_counties error=sd nprocs=10 \
+    where="start_time > '2004-01-01' AND start_time < '2010-01-01'"
```
```suggestion
t.rast.line -l -g input=tempmean zones=comparison_counties error=sd nprocs=10 \
    where="start_time > '2004-01-01' AND start_time < '2010-01-01'"
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.html:
```diff
@@ -0,0 +1,230 @@
+
+<p>
+<div class="code"><pre>
+t.rast.line -l -g input=tempmean@climate_2000_2012 \
+    zones=comparison_counties error=sd nprocs=10 \
+    y_axis_limits=-10,40 y_label="mean temperature"
+    
```
```suggestion
t.rast.line -l -g input=tempmean \
    zones=comparison_counties error=sd nprocs=10 \
    y_axis_limits=-10,40 y_label="mean temperature"
```

veroandreo commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,673 @@
+# % key: dpi
+# % type: integer
+# % label: DPI
+# % description: Resolution of plot.
```
```suggestion
# % description: Plot resolution in DPI.
```

veroandreo reviewed (commented):
I tested the addon, nice! Some suggestions inline. When I pass an nonexistent strds, I get: 
```
ERROR: Dataset <precip_sum@climate_2000_2012> of type <strds> not found in
       temporal database
Traceback (most recent call last):
  File "/home/vandreo/.grass8/addons/scripts/t.rast.line", line 673, in <module>
    sys.exit(main(*gs.parser()))
             ^^^^^^^^^^^^^^^^^^
  File "/home/vandreo/.grass8/addons/scripts/t.rast.line", line 541, in main
    t_info = gs.parse_command("t.info", flags="g", input=options["input"])
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/grass83/etc/python/grass/script/core.py", line 573, in parse_command
    res = read_command(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/grass83/etc/python/grass/script/core.py", line 533, in read_command
    return handle_errors(returncode, stdout, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/grass83/etc/python/grass/script/core.py", line 345, in handle_errors
    raise CalledModuleError(module=module, code=code, returncode=returncode)
grass.exceptions.CalledModuleError: Module run `t.info -g input=precip_sum@climate_2000_2012` ended with an error.
```
Perhaps the error can be caught differently? i.e., to get only the message and not the rest? Is that possible?
 
Also, what are the atexit, gs and sys empty files?

ecodiv commented:
> Also, what are the atexit, gs and sys empty files?

@veroandreo where do you see these empty files?

ecodiv committed:
Apply suggestions from code review

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

ecodiv committed:
Update src/temporal/t.rast.line/t.rast.line.py

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

veroandreo commented:
> > Also, what are the atexit, gs and sys empty files?
> 
> @veroandreo where do you see these empty files?

Files changed tab within this PR

![image](https://github.com/OSGeo/grass-addons/assets/20075188/3ddcde74-c4da-44b7-ab99-4c442d86633b)

ecodiv commented:
> > > Also, what are the atexit, gs and sys empty files?
> > 
> > 
> > @veroandreo where do you see these empty files?
> 
> Files changed tab within this PR
> 
Thanks, result of some previous failed runs. Removed.

ecodiv committed:
Better error handling and added illustration

Illustration of an use case; comparing seasonal patterns in FCover between land cover types.

Use try / exception so when strds does not exists, the module fails immediately, avoiding all the extra lines in error message. This is not an ideal solutions as it now gives two error message, but not sure how to do this otherwise.

Added clean-up function to clean up intermediate layers.

ecodiv committed:
Merge branch 'OSGeo:grass8' into t_rast_line

ecodiv committed:
Merge branch 't_rast_line' of github.com:ecodiv/grass-addons into t_rast_line

petrasovaa commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,700 @@
+#############################################################################
+
+# %module
+# % description: Draws line plots of the raster maps in a space-time raster dataset
+# % keyword: display
+# % keyword: raster
+# % keyword: plot
+# % keyword: line plot
+# % keyword: statistics
+# %end
```
The first keyword should be temporal, since it's in the t* family. For addons it doesn't matter that much, but still.
```suggestion
# % description: Draws line plots of the raster maps in a space-time raster dataset
# % keyword: temporal
# % keyword: display
# % keyword: raster
# % keyword: plot
# % keyword: line plot
# % keyword: statistics
# %end
```

petrasovaa commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,700 @@
+# %end
+
+# %option G_OPT_M_NPROCS
+# % key: nprocs
+# % label: Number of processor threads to use.
+# % answer: 1
```
```suggestion
```

petrasovaa commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,700 @@
+# %end
+
+# %option G_OPT_STRDS_INPUT
+# % description: Input space-time raster dataset.
+# % required: yes
```
```suggestion
```

petrasovaa commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,700 @@
+# %end
+
+# %option G_OPT_T_WHERE
+# % description: WHERE conditions of SQL statement without 'where' keyword used in the temporal GIS framework. Example: start_time > '2001-01-01 12:30:00'
```
```suggestion
```

petrasovaa commented on src/temporal/t.rast.line/t.rast.line.py:
```diff
@@ -0,0 +1,700 @@
+    gs.message("Getting the strds metadata...")
+    try:
+        t_info = gs.parse_command("t.info", flags="g", input=options["input"])
+    except:
```
It's not a good practice to use except without the exception type, you can use grass.exceptions.CalledModuleError here.

ecodiv replied:
Thanks, implemented

petrasovaa commented:
> > > > Also, what are the atexit, gs and sys empty files?
> > > 
> > > 
> > > @veroandreo where do you see these empty files?
> > 
> > 
> > Files changed tab within this PR
> 
> Thanks, result of some previous failed runs. Removed.

It's still there...

petrasovaa commented:
> I tested the addon, nice! Some suggestions inline. When I pass an nonexistent strds, I get:
> 
> ```
> ERROR: Dataset <precip_sum@climate_2000_2012> of type <strds> not found in
>        temporal database
> Traceback (most recent call last):
>   File "/home/vandreo/.grass8/addons/scripts/t.rast.line", line 673, in <module>
>     sys.exit(main(*gs.parser()))
>              ^^^^^^^^^^^^^^^^^^
>   File "/home/vandreo/.grass8/addons/scripts/t.rast.line", line 541, in main
>     t_info = gs.parse_command("t.info", flags="g", input=options["input"])
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/grass83/etc/python/grass/script/core.py", line 573, in parse_command
>     res = read_command(*args, **kwargs)
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/grass83/etc/python/grass/script/core.py", line 533, in read_command
>     return handle_errors(returncode, stdout, args, kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib/grass83/etc/python/grass/script/core.py", line 345, in handle_errors
>     raise CalledModuleError(module=module, code=code, returncode=returncode)
> grass.exceptions.CalledModuleError: Module run `t.info -g input=precip_sum@climate_2000_2012` ended with an error.
> ```
> 
> Perhaps the error can be caught differently? i.e., to get only the message and not the rest? Is that possible?
> 
You can pass `errors="fatal"` into run_command, that should do it:
https://grass.osgeo.org/grass83/manuals/libpython/script.html#script.core.handle_errors

petrasovaa reviewed (commented):
Some messages are translatable, some not, so if you could make them all translatable, that would be better.

ecodiv commented:
> Some messages are translatable, some not, so if you could make them all translatable, that would be better. 

@petrasovaa  I normally do that, but for some reason, if I run it with e.g., 

```
# Get strds type
gs.message(_("Getting the strds metadata..."))
```

I'm getting an error message about _ being references before assignment. 

```
GRASS WGS84/VITO_FCover:~ > t.rast.line input=FCOVER300 zones=LUCesri3 nprocs=10
Traceback (most recent call last):
  File "/home/paulo/.grass8/addons/scripts/t.rast.line", line 694, in <module>
    sys.exit(main(*gs.parser()))
  File "/home/paulo/.grass8/addons/scripts/t.rast.line", line 557, in main
    gs.message(_("Getting the strds metadata..."))
UnboundLocalError: local variable '_' referenced before assignment
```
Any idea what I am doing wrong here?

ecodiv commented:
> > > > > Also, what are the atexit, gs and sys empty files?
> > > > 
> > > > 
> > > > @veroandreo where do you see these empty files?
> > > 
> > > 
> > > Files changed tab within this PR
> > 
> > 
> > Thanks, result of some previous failed runs. Removed.
> 
> It's still there...

Ah, in the src folder. I really do not understand how they end up there. But in any case, I'll remove them. Thanks for pointing out.

ecodiv committed:
Apply suggestions from code review

Co-authored-by: Anna Petrasova <kratochanna@gmail.com>

ecodiv commented:
> `errors="fatal"`

That also results in a double error message. Adding

```
except CalledModuleEror:
    return False
``` 

Does the trick, it seems.

ecodiv committed:
further improvements based on reviewers suggestions

petrasovaa commented:
> > `errors="fatal"`
> 
> That also results in a double error message. Adding
> 
Strange, I tested with:
```
gs.read_command("r.info", map="aa", errors="fatal")
```
and I get
```
ERROR: Raster map <aa> not found
ERROR: Module r.info (r.info map=aa) failed with non-zero return code 1
```

petrasovaa commented:
> > Some messages are translatable, some not, so if you could make them all translatable, that would be better.
> 
> @petrasovaa I normally do that, but for some reason, if I run it with e.g.,
> 
> ```
> # Get strds type
> gs.message(_("Getting the strds metadata..."))
> ```
> 
> I'm getting an error message about _ being references before assignment.
> 
> ```
> GRASS WGS84/VITO_FCover:~ > t.rast.line input=FCOVER300 zones=LUCesri3 nprocs=10
> Traceback (most recent call last):
>   File "/home/paulo/.grass8/addons/scripts/t.rast.line", line 694, in <module>
>     sys.exit(main(*gs.parser()))
>   File "/home/paulo/.grass8/addons/scripts/t.rast.line", line 557, in main
>     gs.message(_("Getting the strds metadata..."))
> UnboundLocalError: local variable '_' referenced before assignment
> ```
> 
> Any idea what I am doing wrong here?

Remove the `_` from `for i, _ in enumerate(cats_ids):` (and one other place), that should fix it.

ecodiv committed:
better error handling for maps without color table

ecodiv committed:
remove redundant f-string
